### ISSUE TITLE
disallowAnonymousFunctions: New rule value to allow anonymous callbacks

### ISSUE THREAD

**Issue by benjamincharity** [jscs-dev/node-jscs#1562]
Just getting set up with JSCS for the first time and can't quite seem to figure this out. 

I want to require that functions are named *except* when used in a promise. Showing might be easier than explaining...

```
// bad
var foo = function() {};

// good
function foo() {};
```

But since our codebase heavily uses promises, something like this needs to **not** throw an error:

```
MyService.doThing(param).then(function(result) {
    // stuff..
});
```

Is there a rule or set of rules that will

- a) Stop us from creating unnamed functions (assigned or unassigned)
- b) Still allow unnamed functions inside `then()` statements

Thanks in advance!

**Comment by benjamincharity** [comment]
Also, I've noticed that even with `requireLineFeedAtFileEnd` set to `true`, no error is surfaced when there is no empty line at the end of my JS file. Am I misinterpreting this rule?

```
// This should throw an error

    // file contents above...
}); // end of file with no blank line
```

```
// This should be valid

    // file contents above...
});
    // empty line here to end file
```

**Comment by hzoo** [comment]
Thanks for contributing! - In http://jscs.info/rules.html there's http://jscs.info/rule/disallowAnonymousFunctions.html although it would error for all of them.

In your example

```js
// bad
var foo = function() {};

// good
function foo() {};
```
the good example would be a `FunctionDeclaration`

```js
var foo = function() {}; // function expression (no name)
var foo = function foo2() {}; // named function expression
```

So I think you want function declarations to be used but still allow callbacks?

Also if you have multiple issues I would make a separate one to help with discussion.

**Comment by benjamincharity** [comment]
Yes I believe you are correct. It just seems that those two desires fight against each other. Requiring function declarations throws errors for my anonymous functions inside callbacks.

Solvable, but going through our entire codebase and naming all functions is a bit too much work at the moment (we have a very promise-heavy codebase).

**Comment by mrjoelkemp** [comment]
Sounds like a new rule value `allExcept: ['callbacks']` would do it.

@benjamincharity Would you like to give the implementation of this new value a shot? We could use the help. If you have any questions, feel free to ask questions on this thread or on [gitter](https://gitter.im/jscs-dev/node-jscs).

**Comment by benjamincharity** [comment]
Sure thing. I'll take a stab in the next few days when I get some time.

**Comment by hzoo** [comment]
We released [3.0](https://github.com/jscs-dev/node-jscs/releases/tag/v3.0.0) recently! We also wrote a [post](https://medium.com/@markelog/jscs-end-of-the-line-bc9bf0b3fdb2#.oot3wvgpu) about how 3.0 will be our last major release (since we are deprecating the project) and that we are [merging with ESLint](http://eslint.org/blog/2016/04/welcoming-jscs-to-eslint)!

Thus we won't be working on any new features moving forward which includes new rules/options. We will continue to fix bugs and work on CST/better autofixing for current rules in the upcoming months both to help JSCS users and to gain experience that we can hopefully use in ESLint. But otherwise we both teams will be working on getting [ESLint feature parity with JSCS](https://github.com/eslint/eslint/milestones/JSCS%20Compatibility) and just a nice transition for our users.

I'm putting an `orphaned` label to note this issue isn't planned to be worked on, but might be a good contribution as a custom rule (not in JSCS core itself).

I would encourage anyone who wants the rule to try writing it themselves: I would either write a custom rule or submit an issue to ESLint and if accepted a PR. Otherwise a custom rule that won't be in core repo itself.

JSCS rules are also very similar to ESLint rules: here are some resources to get started: 

- [Project-specific Lint Rules with ESLint](https://gist.github.com/jareware/7179093)
- [Creating Custom Rules for ESLint](https://insideops.wordpress.com/2015/12/08/creating-custom-rules-for-eslint)
- [Maintaining JavaScript Code Quality with ESLint ](https://www.paypal-engineering.com/2014/12/12/maintaining-javascript-code-quality-with-eslint/)
